### PR TITLE
[[ Bug 19967 ]] Add `send script <script> to <obj>` syntax

### PR DESCRIPTION
**Pull request by livecodeali**
The current send command attempts to evaluate parameters in the
current context before sending the message. In some cases this
is not desirable, and furthermore there are issues with the
evaluation of multiple parameters using the send command.

Since we can't break this backwards compatibility with the send
command, this patch adds a `send script` form of the command which
simply forwards the script to be sent unmodified to the target
object to be executed, using MCObject::domess.

### CODE REVIEW

File: tests/lcs/core/engine/send.livecodescript [hunk 1]
```diff
@@ -33,6 +33,27 @@ on TestSendParams
...
+*/
+on TestSendLegacy
+	local tStack
+	create stack
+	put it into tStack
+	set the defaultStack to the short name of tStack
+	create button
+	
+	local tSendScript
+	put "delete button 1 of me" into tSendScript
```
Comment (montegoulding) [inline tests/lcs/core/engine/send.livecodescript]: @livecodeali should we have a test that does parameter evaluation in the calling script context?

File: engine/src/cmdse.cpp [hunk 2]
```diff
@@ -692,7 +684,19 @@ Parse_stat MCMessage::parse(MCScriptPoint &sp)
...
 			return PS_ERROR;
 		}
-		return gettime(sp, &in, units);
+        
+        if (sp.skip_token(SP_FACTOR, TT_PREP, PT_AS) == PS_NORMAL)
+        {
+            if (!send ||
+                sp.skip_token(SP_FACTOR, TT_PROPERTY, P_SCRIPT) != PS_NORMAL)
+            {
+                MCperror->add(PE_SEND_BADAS, sp);
```
Comment (montegoulding) [inline engine/src/cmdse.cpp]: @livecodeali is it bad that I want you to change the error to `PE_SEND_BADASS` :-P

File: engine/src/cmdse.cpp [hunk 3]
```diff
@@ -692,7 +684,19 @@ Parse_stat MCMessage::parse(MCScriptPoint &sp)
...
+                return PS_ERROR;
+            }
+            script = True;
+        }
+        else
+            return gettime(sp, &(&in), units);
```
Comment (montegoulding) [inline engine/src/cmdse.cpp]: @livecodeali do we want to support send in time as script? I guess we don't have a use case for it at the moment.

File: tests/lcs/core/engine/send.livecodescript [hunk 4]
```diff
@@ -71,4 +92,19 @@ on TestSendDeleteMe
...
+
+on TestSendAsScript
+	local tStack
+	create stack
+	put it into tStack
+	set the defaultStack to the short name of tStack
+	create button
+	
+	local tSendScript
+	put "delete button 1 of me" into tSendScript
```
Comment (montegoulding) [inline tests/lcs/core/engine/send.livecodescript:108]: @livecodeali should we test param evaluation in the object context? Maybe getting the value of a script local? Hmm... if we can do that we could add script locals to the SE when not debugging. That would be lovely.

**Comment by runrevmark** [comment]
I wonder if:
```
   send script ... to ...
```
Might be a better choice?

It is more aligned with dispatch (which is 'dispatch command | function'), it also avoids the use of 'as' (which we need to be careful with if we ever want it to be a type conversion operator), and means that the right hand side of 'send' isn't overloaded (as it is target + delay - albeit delay isn't implemented in this case).

**Comment by livecodeali** [comment]
@runrevmark The reason I didn't do that was because of the possibility of
```
on script
// Do something
end script
```
Although I suppose that doesn't strictly make a difference

File: tests/lcs/core/engine/send.livecodescript (hunk 1 above)
Comment (livecodeali) [reply tests/lcs/core/engine/send.livecodescript]: oops, that's actually what this was meant to do!

File: engine/src/cmdse.cpp (hunk 3 above)
Comment (livecodeali) [reply engine/src/cmdse.cpp]: Yeah, I'd say not right now, unless there's a problem it solves nicely

**Comment by runrevmark** [comment]
@livecodeali: I'm not sure that's a problem:
```
  send script to the selobj
```
is distinct from:
```
  send script <expr> to the selobj
```
The syntax of send IIRC is:
```
  'send' <expr> to <container> [ in <time> ]
```
So, the proposed addition would make it additionally parse:
```
  'send' 'script' <expr> to <container>
```

**Comment by livecodeali** [comment]
@runrevmark Yeah that's what I meant by 'it doesn't strictly make a difference', I should have said 'it's actually not a problem at all' :-D

**Comment by runrevmark** [comment]
@livecodeali I'm not sure you did say 'it doesn't strictly make a difference' ;)

**Comment by montegoulding** [comment]
Hmm... what about unquoted literal `script`. This currently works:
```
on script
   answer "foo"
end script

on mouseUp
   send script to me
end mouseUp
```

**Comment by montegoulding** [comment]
I guess you could code around that though...

**Comment by livecodeali** [comment]
@montegoulding That was what I meant, but it's still not ambiguous because you need the `to`

**Comment by runrevmark** [comment]
In this case there is no ambiguity because send requires a single expression after 'send' (regardless of whatever comes after). In this case, we require 'script' *then* an expression. It's a good example of why it is never a good idea to allow an expression to be two expressions immediately following each other (e.g. in C you can do "foo" "bar" - although only for string literals, I think).

**Comment by mwieder** [comment]
I'm quite distressed that you can create a handler named 'script'. Shouldn't that be a reserved keyword? I can't see anything but confusion coming from this.

**Comment by livecodeali** [comment]
Only commands and functions are reserved for command and function names.

**Comment by livecodeali** [comment]
Actually it's not true that only those are ;-) But property names aren't reserved for command and function names as there's (technically) no ambiguity

**Comment by mwieder** [comment]
Interesting. Thanks for the explanation. I'm still not tempted to go down that rabbit hole, though.

**Comment by livecodeali** [comment]
Nor should you! Anyway, nothing to see here on that front, we were just considering potential backwards compatibility issues :-)

**Comment by mwieder** [comment]
Right. I'm on board with the rest of this, although @runrevmark's comment about 'in time' not being supported in this case is disturbing. Is there a reason not to also add MCEngineExecSendScriptInTime?

**Comment by runrevmark** [comment]
@mwieder: Yes - it adds complexity to the implementation, and isn't needed for the problem we are currently trying to solve (i.e. fixing the message box properly). Things are never 'just about syntax' - the infrastructure to support the syntax has to be there too.

At the moment, only handler calls can be processed as pending messages - not script fragments (due to the way pending messages are implemented). This is basically a replacement for a 'correct' 'do in object' which we can't do at the moment because of 'do in widget' (for the browser) which makes it ambiguous.

**Comment by mwieder** [comment]
Ah. Got it. I guess the loss of orthogonality is more than offset by the complexity of reworking the infrastructure.

**Comment by livecodeali** [comment]
Plus there's the fact that this is a release branch fix - it's already more complicated than is ideal for merging into what will be an RC 3. We may be able to add the ...in time version, but it'll only get done if there is a compelling use-case and even then will be develop only!

**Comment by runrevmark** [comment]
So, having pondered this a bit - the rather old 'crufty' code in Send would appear to have been trying to emulate executing a closure created in the calling context in the context of the target object. Of course, one issue here is that 'context of the target object' is a little ill-defined at present - however, I think it basically has been notionally intended to mean that 'me' is the target object, direct handler invocations use the public handlers of the target (although I'm not sure that works with the existing 'new handlerlist' approach - would need to check), and 'this' things are also the target context (following the rules of a send, rather than a call).

Originally send would do:
- attempt to parse the string as "cmd paramvar, ..., paramvar", and if it managed it converted it to a handler invocation.
- just use the string poked into a temporary handler as script, and then called using 'domess'.

If the variable context of the calling handler is captured in the caller, then these two things are equivalent - however, at the moment only a vague attempt is made at capturing context in the first case (i.e. it does because it evaluates the variables in the caller - which is the same as capturing them and then just evaluating them in the callee).

So, 'send script' is useful for now to disambiguate the case we are encountering in the message box. However, if we could figure out how to capture caller context and then execute the code in the target context with said captured context (this is not the same as allowing access to private things in the target object - its variable context shouldn't be accessible at all) then the send forms become equivalent.

TL;DR - The need for the 'script' token before the first expression is only needed *now* because we do not capture vars. When we do, it can become an optional keyword which has no effect. (Filed as http://quality.livecode.com/show_bug.cgi?id=19986).

File: engine/src/cmdse.cpp [hunk 5]
```diff
@@ -642,37 +642,45 @@ void MCDispatchCmd::exec_ctxt(MCExecContext &ctxt)
...
 {
 	initpoint(sp);
 
-	if (sp.parseexp(False, True, &message) != PS_NORMAL)
-	{
-		MCperror->add(PE_SEND_BADEXP, sp);
-		return PS_ERROR;
-	}
+    MCScriptPoint oldsp(sp);
+    if (send && sp.skip_token(SP_FACTOR, TT_PROPERTY, P_SCRIPT) == PS_NORMAL)
```
Comment (runrevmark) [inline engine/src/cmdse.cpp:650]: Can this not be:
```
if (send && sp.skip_token(...) == PS_NORMAL) script = True;
if (sp.parseexpr(False, True, &(&message)) != ...) ...
```

Comment (livecodeali) [reply engine/src/cmdse.cpp:650]: I don't think so, because the `message` might be `script` as UQL

Comment (runrevmark) [reply engine/src/cmdse.cpp:650]: Ah! Got it :)

**Comment by runrevmark** [comment]
@livecode-vulcan review ok d3c5eb7

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok d3c5eb78ed40418c55553c5c920a7be23669eb85

**Comment by livecode-vulcan** [comment]
😎 test success d3c5eb78ed40418c55553c5c920a7be23669eb85

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/454)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/462)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/441)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/384)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/384)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/449)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/454)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/452)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/465)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/452)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/461)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/479)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/448)
